### PR TITLE
Fix some flaky reporting corner cases

### LINKED ISSUE #4183: Bizarre quirk with how Hypothesis formats the error message for falsifying examples
Consider this test:
```python
counter = 0

@hypothesis.given(dummy=st.text())
def test_that_prints_rich_error(dummy: str):
    global counter
    counter += 1
    if counter % 5 == 0:
        raise ExceptionGroup(
            "outer exceptiongroup", [ZeroDivisionError("this is a fake error")]
        )
```
It will fail nondeterministically, of course. Hypothesis reports a rich error message describing the failure:
```
...
  |   File "/opt/homebrew/Caskroom/miniforge/base/envs/py311/lib/python3.11/site-packages/hypothesis/core.py", line 1722, in wrapped_test
  |     raise the_error_hypothesis_found
  | hypothesis.errors.FlakyFailure: Hypothesis test_that_prints_rich_error(dummy='\x02\x1d\U000c4bc0') produces unreliable results: Falsified on the first call but did not on a subsequent one (1 sub-exception)
  | Falsifying example: test_that_prints_rich_error(
  |     dummy='\x02\x1d\U000c4bc0',
  | )
  | Failed to reproduce exception. Expected:
  | + Exception Group Traceback (most recent call last):
  |   |   File "/Users/tmaxwell/code/anthropic/belt/tests/test_hypothesis_utils.py", line 399, in test_that_prints_rich_error
  |   |     raise ExceptionGroup(
  |   | ExceptionGroup: outer exceptiongroup (1 sub-exception)
  |   +-+---------------- 1 ----------------
  |     | ZeroDivisionError: this is a fake error
  |     +------------------------------------
  |
  | Explanation:
  |     These lines were always and only run by failing examples:
  |         /Users/tmaxwell/code/anthropic/belt/tests/test_hypothesis_utils.py:399
  +-+---------------- 1 ----------------
    | Exception Group Traceback (most recent call last):
    ...
```

Now consider this one:
```python
counter = 0

@hypothesis.given(dummy=st.text())
def test_that_does_not_print_rich_error(dummy: str):
    global counter
    counter += 1
    if counter % 5 == 0:
        try:
            raise ZeroDivisionError("this is a fake error")
        except* Exception as e:
            raise
```
This also fails nondeterministically, but with a much less helpful message:
```
...
  |   File "/opt/homebrew/Caskroom/miniforge/base/envs/py311/lib/python3.11/site-packages/hypothesis/core.py", line 1722, in wrapped_test
  |     raise the_error_hypothesis_found
  | ExceptionGroup:  (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/tmaxwell/code/anthropic/belt/tests/test_hypothesis_utils.py", line 389, in test_that_does_not_print_rich_error
    |     raise ZeroDivisionError("this is a fake error")
    | ZeroDivisionError: this is a fake error
    +------------------------------------
```

What is the difference...? I feel like I'm going insane.

This is with Hypothesis v6.112.1.

### LINKED ISSUE #4228: Avoid raising FlakyReplay to users when multiple bugs are found
I'm trying to track down the cause of a flaky test. However, sometimes the test fails with FlakyFailure and other times with FlakyReplay. These are both errors, but no reproduce_failure hash is provided when the error is FlakyReplay.

### What is the difference between a FlakyReplay and a FlakyFailure?

```python
class Flaky(_Trimmable):
    """Base class for indeterministic failures. Usually one of the more
    specific subclasses (FlakyFailure or FlakyStrategyDefinition) is raised."""


class FlakyReplay(Flaky):
    """Internal error raised by the conjecture engine if flaky failures are
    detected during replay.

    Carries information allowing the runner to reconstruct the flakiness as
    a FlakyFailure exception group for final presentation.
    """

    def __init__(self, reason, interesting_origins=None):
        super().__init__(reason)
        self.reason = reason
        self._interesting_origins = interesting_origins

class FlakyFailure(ExceptionGroup, Flaky):
    """This function appears to fail non-deterministically: We have seen it
    fail when passed this example at least once, but a subsequent invocation
    did not fail, or caused a distinct error.

    Common causes for this problem are:
        1. The function depends on external state. e.g. it uses an external
           random number generator. Try to make a version that passes all the
           relevant state in from Hypothesis.
        2. The function is suffering from too much recursion and its failure
           depends sensitively on where it's been called from.
        3. The function is timing sensitive and can fail or pass depending on
           how long it takes. Try breaking it up into smaller functions which
           don't do that and testing those instead.
    """

    def __new__(cls, msg, group):
        # The Exception mixin forces this an ExceptionGroup (only accepting
        # Exceptions, not BaseException). Usually BaseException is raised
        # directly and will hence not be part of a FlakyFailure, but I'm not
        # sure this assumption holds everywhere. So wrap any BaseExceptions.
        group = list(group)
        for i, exc in enumerate(group):
            if not isinstance(exc, Exception):
                err = _WrappedBaseException()
                err.__cause__ = err.__context__ = exc
                group[i] = err
        return ExceptionGroup.__new__(cls, msg, group)
```
Above is the relevant source code from hypothesis. My understanding is that FlakyReplay is "internal" which means it should not be shown to the user, so something is wrong if I am seeing this error in my logs as the final error? In my case, this error is raised when the status goes from INTERESTING to INTERESTING. (Is there documentation for the meaning of this status? [Couldn't find any mention in the source.](https://github.com/search?q=repo%3AHypothesisWorks%2Fhypothesis%20Status.INTERESTING&type=code))

### Why is reproduce_failure provided for one and not the other?

Is this because FlakyReplay is not intended to be raise to the user?

### PR DESCRIPTION
**Pull request by jobh**
Closes #4183 
Closes #4228 

Two corner cases in reporting of flaky errors.

One of them is a straightforward missing handling of flakiness at `mark_interesting` time.

The other deals with a surprising aspect of how a naked exception is wrapped in an `ExceptionGroup` when caught and reraised by `except*` — the group's traceback points not to where the naked exception is raised, nor to the location of the `except*` block, but instead to the _caller_ one frame above the `except*`:
```
def wrapped():
    try:
        raise Exception  # innermost frame of EceptionGroup.exceptions[0]
    except* Exception:
        raise

def naked():
    raise Exception  # innermost frame of Exception

if __name__ == "__main__":
    import traceback
    try:
        naked()
    except Exception as e:
        print("naked:", traceback.format_tb(e.__traceback__)[-1])
    try:
        wrapped()  # innermost frame of ExceptionGroup
    except Exception as e:
        print("wrapper:", traceback.format_tb(e.__traceback__)[-1])
        print("wrapped:", traceback.format_tb(e.exceptions[0].__traceback__)[-1])
```

I'm not sure if this is consistent between python implementations, so the final commit adds a check. Will revert once CI results are in.

### CODE REVIEW

File: hypothesis-python/src/hypothesis/core.py [hunk 1]
```diff
@@ -1119,7 +1123,14 @@ def _execute_once_for_engine(self, data: ConjectureData) -> None:
...
             # Hypothesis-internal code, re-raise it as a fatal error instead
             # of treating it as a test failure.
-            filepath = traceback.extract_tb(e.__traceback__)[-1][0]
+            if isinstance(e, BaseExceptionGroup) and len(e.exceptions) == 1:
+                # When a naked exception is implicitly wrapped in an ExceptionGroup
+                # due to a re-raising "except*", the ExceptionGroup is constructed
+                # in the caller stack frame (see #4183).
+                tb = e.exceptions[0].__traceback__
+            else:
+                tb = e.__traceback__
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/core.py:1132]: What if there are multiple layers, e.g. `ExceptionGroup("outer", ExceptionGroup("inner", [ValueError()]))`?  Should we use `while` to unwrap them all?

(I'd appreciate a code comment about this so we're less baffled when revisiting in a few years for something even more complicated 😅)

**Review by Zac-HD — COMMENTED** [review]
Thanks Joachim!  This looks great and I'm sure will be increasingly useful over time 😁

File: hypothesis-python/src/hypothesis/core.py (hunk 1 above)
Comment (jobh) [reply hypothesis-python/src/hypothesis/core.py]: Yep done.

The "missing stack frame" on implicit wrapping of naked exceptions in `except*` — I won't call it a cpython bug, as it doesn't contradict any documentation I've seen, but IMO it's surprising behaviour. Do you think it would be worth raising a concern upstream @Zac-HD ?

**Comment by tybug** [comment]
Test failures are at least in part my fault - I'll look into this and come back with a fix.

> FAILED tests/conjecture/test_shrinker.py::test_redistribute_numeric_pairs - KeyError: 'stopped-because'

**Comment by tybug** [comment]
#4240 should address failures, and in the meantime we can retry jobs that fail because of it.

File: hypothesis-python/src/hypothesis/core.py (hunk 1 above)
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/core.py]: Yeah, I think it's worth opening an issue to say something like "we were surprised by this behavior in $PR, and think that $proposal would be less surprising and more useful.  If you agree, up to you whether this treated as a bugfix, or a minor change for 3.14"

**Review by Zac-HD — APPROVED** [review]
Thanks again, all!

**Comment by carterbox** [comment]
Thanks, @jobh for patching this quickly!

File: hypothesis-python/src/hypothesis/core.py (hunk 1 above)
Comment (jobh) [reply hypothesis-python/src/hypothesis/core.py]: Too late for a code comment in this PR, but FYI @Zac-HD : Cpython will have this fixed in 3.12+, so the workaround can be removed once 3.11 is EOL.

Comment (tybug) [reply hypothesis-python/src/hypothesis/core.py]: Nice! I've stuck a note to that effect onto #4252.
